### DISCUSSION TITLE
in `vertico-grid-mode`, `vertico-grid-annotate` cause grids not align well

### DISCUSSION THREAD

**Discussion by eval-exec** [minad/vertico#544] [Q&A]
Hello, Thanks for your vertico project, it helped me alot.

When I enable `vertico-grid-mode` for `M-x` command, the grids not align well.

Minimal config to reproduce:
![image](https://github.com/user-attachments/assets/da1a0c84-14f7-435a-a39d-652fc9896b9f)

```elisp
(require 'package)
(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/") t)
;; Comment/uncomment this line to enable MELPA Stable if desired.  See `package-archive-priorities`
;; and `package-pinned-packages`. Most users will not need or want to do this.
;;(add-to-list 'package-archives '("melpa-stable" . "https://stable.melpa.org/packages/") t)
(package-initialize)

(setq-default use-package-always-ensure t)



(use-package marginalia
  :hook
  (vertico-mode . marginalia-mode)
  )

(use-package vertico
  :config
  (vertico-mode)
  (setq vertico-grid-annotate 10)
  (setq vertico-multiform-commands
	'(
	   (execute-extended-command grid)
	   )
	)
  (vertico-multiform-mode))
```

**Comment by eval-exec** [comment]
Hello, @minad . How to let these grids align well? Thank you.

**Reply by minad** [reply]
The annotation alignment is coming from Marginalia and is not compatible with Vertico Grid alignment. Unfortunately this is a limitation of the `annotation-function` and `affixation-function` API of `completing-read`. A better annotation API would be needed - I call it a `decoration-function`, where the decoration function just returns a list of columns with some meta data (min/max column width, desired alignment, prefix/suffix, left/right padding, left/right truncation) which are then completely aligned by the UI. I have mentioned this multiple times before in various threads and on mailing lists, but unfortunately this information is not easy to find.

The only annotations which are supported by grid mode are short unaligned annotations. See for example https://github.com/minad/vertico/wiki#annotate-m-x-commands-with-keybindings-in-flatunobtrusive-mode. I use key binding annotations in flat/grid more. Verbose annotations like docstring are not ideal for grid mode anyway, since grid mode is supposed to show as many candidates as possible in a dense way.

**Reply by eval-exec** [reply]
Thank you, I want to hack `marginalia--align`, how to let `marginalia--align` know wheather the `cand` is in first column or second column or third column?
https://github.com/minad/marginalia/blob/643a5f50c9f9d0698c8b1d72678886aefcf69052/marginalia.el#L1223-L1247

**Reply by eval-exec** [reply]
I hacked `vertico--arrange-candidates` by this way: https://github.com/minad/vertico/pull/543/files
now the first column align well.
![image](https://github.com/user-attachments/assets/da71f449-3802-4664-9e20-fcddce5955b9)

but second column not align well. I think it's because marginalia--align-to not set right `:align-to` for second column's cand: https://github.com/minad/marginalia/blob/643a5f50c9f9d0698c8b1d72678886aefcf69052/marginalia.el#L1239

Look above picture, `elisp/embark.el` not align well, and I tracked marginalia--affixate, it show this:
```elisp
("elisp/embark.el" #1# #(" │-rw-r--r--│   1.6k│2024-12-11 Wed 20:15:23│" 0 1 (display (space :align-to (+ left 33)) marginalia--align t) 2 3 (face marginalia-file-priv-no) 3 4 (face marginalia-file-priv-read) 4 5 (face marginalia-file-priv-write) 5 6 (face marginalia-file-priv-no) 6 7 (face marginalia-file-priv-read) 7 9 (face marginalia-file-priv-no) 9 10 (face marginalia-file-priv-read) 10 12 (face marginalia-file-priv-no) 13 15 (face marginalia-size) 15 20 (help-echo "1599" face marginalia-size) 21 44 (help-echo "2024-12-11 20:15:23" face marginalia-date)))
```
I think the `(space :align-to (+ left 33)` is not suitable for second column
should hacked the `(+ left 33)` to `(+ left_column_width 33)`

**Reply by eval-exec** [reply]
Finaly: I hacked `vertico--arrange-candidates`:
```elisp
(cl-loop for row from 0 to (1- (min vertico-count vertico--total)) collect
          (let ((line (list "\n")))
            (cl-loop for col from (1- vertico-grid--columns) downto 0 do
              (when-let ((cand (nth (+ row (* col vertico-count)) cands)))
                (push cand line)
				(progn
				  (let* ((sep-index
						   (string-search marginalia-separator cand))
						  (t-property
							(text-properties-at (1- sep-index ) cand)))
					;; (remove-text-properties (1- sep-index) sep-index '(display) cand)
					(when (> col 0)
					  (set-text-properties
						(1- sep-index)
						sep-index
						'(display
						   (marginalia--align t display (space :align-to (+ left (+ 80 33))))
						   ) cand))
					)
				  )
                (when (> col 0)
                  (push vertico-grid-separator line)
                  (push (propertize " " 'display
                          `(space :align-to (+ left ,(aref width (1- col))))) line))))
            (string-join line)))
```

Then I can align vertico-grid well like this:

![image](https://github.com/user-attachments/assets/03dd9b43-062b-4660-a355-ea1188e9f242)
